modules/OFConnectionManager2/module/src/bundle.c

```diff
@@ -500,52 +517,51 @@ bundle_task(void *cookie)
-                              state->cur_subbundle,
-                              state->cur_offset, state->cur_offset_is_pending);
-                if (state->cur_offset_is_pending == false) {
-                    /* The task is ok to process message at cur_offset.
-                     * The cur_offset is in clear status. */
-                    of_object_t *obj = subbundle->objs[state->cur_offset];
-                    if (obj) {
-                        indigo_error_t rv;
-                        rv = ind_cxn_process_message(cxn, obj);
-                        if (rv == INDIGO_ERROR_PENDING) {
-                            /* The message is an async operation
-                             * Mark the messsage at cur_offset as pending.
-                             * The pending status will be cleared in the next
-                             * runnable cycle.
-                             * The lower layer decides to block the bundle loop.
-                             */
-                            state->cur_offset_is_pending = true;
-                            AIM_LOG_TRACE("bundle_task cur_offset=%u pending\n",
-                                          state->cur_offset);
-                            return IND_SOC_TASK_CONTINUE;
-                        }
+                AIM_LOG_TRACE("cur_subbundle=%u cur_offset=%u is_paused=%d",
+                              state->cur_subbundle, state->cur_offset,
+                              state->cur_subbundle_is_paused);
+
+                /* The task is ok to process message at cur_offset.
+                 * The cur_offset is in clear status. */
+                of_object_t *obj = subbundle->objs[state->cur_offset];
+                if (obj) {
+                    indigo_error_t rv;
+                    rv = ind_cxn_process_message(cxn, obj);
+                    if (rv == INDIGO_ERROR_PENDING) {
+                        /* The message is an async operation. Set pause flag. */
+                        state->cur_subbundle_is_paused = true; 
+                        AIM_LOG_TRACE("bundle_task cur_offset=%u pending\n",
+                                      state->cur_offset);
                     }
-                } else {
-                    /* This subbundle msg is done. Clear its pending status.
-                     * Fall through to free it.
+                    /* rv is OK, ERROR, or CONTINUE, then allow next message.
+                     * The obj should have been duplicated in the OFStateManager
+                     * or driver. It is safe to free the obj.
+                     * TODO: special case if this obj is the last message in
+                     *       subbundle and rv is CONTINUE, we still treat it as
+                     *       PENDING case.
+                     *       the ind_cxn_process_message() should add a new
+                     *       parameter to tell OFStatemanager/driver no more
+                     *       message in this subbundle. 
```

Comment:
sorry, i missed why ind_cxn_process_message needs to tell OFStatemanager/driver there are no more messages.  it looks like we are processing one subbundle at a time until we get a PENDING from the driver.  we can keep going and handle all messages in the subbundle and just wait until all the async operations are completed, and then move on to the next subbundle.  i don't know why we need to tell OFStatemanager or the driver anything.  can you clarify?

Reply:
On the driver or OFStateaManager doesn't know that it is the end of subbundle, the driver may still wait for more message(s) and process them together. Therefore, adding a new parameter will be easier way to let driver or OFStatemanager know "that's it".